### PR TITLE
Compliance test suite

### LINKED ISSUE #557: Create compliance suite for implementations of the package spec
Create a compliance suite that can be used to define expectations on implementations of the package spec. It must answer the question _Does version X of implementation Y support package spec Z?_
We can think on this as a tool to validate implementations in general, but we should target Kibana.

Its main purpose will be to test that a given build of Kibana properly supports a version of the package spec.

It should be possible to define expectations in a BDD-like fashion, so we could have for example:
```Gherkin
Feature: Runtime fields
  Configure fields in a package as runtime fields.

Scenario: Installer leverages runtime parameter
  Given "Installer" is candidate to support "2.8.0".
   When "package-with-runtime-fields" is installed.
   Then index template for "package-with-runtime-fields" includes runtime fields.
```
It should be possible to run these tests against Kibana endpoints, that would represent the role of "Installer".

These tests should reside within the package spec, so features and their expectations are implemented together, but it should be possible to easily run them in other repositories. For example to make them part of Kibana test suites.

As part of this task we should have the framework and an initial set of tests, with a focus on most recent features. It is not expected to have a compliance suite for the whole spec yet.

### PR DESCRIPTION
**Pull request by jsoriano**
Define an initial compliance suite for the Package Spec, to evaluate if a given stack implements a given.

The compliance suite is defined in the `compliance` directory. Tests are defined in Gherkin syntax (aka Cucumber), and executed using Godog. Features can be found in the [`compliance/features`](https://github.com/jsoriano/package-spec/tree/compliance-suite/compliance/features) directory. Each scenario for each feature is tagged with the package spec version that it requires to work.

Step definitions can be found in [`compliance_test.go`](https://github.com/jsoriano/package-spec/blob/compliance-suite/compliance/compliance_test.go). Additional Go code is used to initialize Kibana and Elasticsearch clients. There is also a wrapper for `elastic-package`, used for build and installation of packages. 

Reference packages used on each scenario can be found in [`compliance/testdata/packages`](https://github.com/jsoriano/package-spec/tree/compliance-suite/compliance/testdata/packages).

Compliance suite can be executed now with:
```
TEST_SPEC_VERSION=2.9.0 make -C compliance/ test
```

Execution can be customized with environment variables:
* `TEST_SPEC_FEATURES`, with a comma-separated list of paths relative to `compliance` with feature files to execute.

Stack to test for compliance can be configured with the following environment variables. Equivalent variables from `elastic-package stack shellinit` are also supported:
* `ELASTICSEARCH_HOST`
* `ELASTICSEARCH_USERNAME`
* `ELASTICSEARCH_PASSWORD`
* `KIBANA_HOST`.

It works with compose and serverless stacks managed by elastic-package.

Closes #557.

### CODE REVIEW

**Review by mrodm — APPROVED** [review]
Great job!!
This kind of tests look easy to define.

I've tested locally to run the same as in CI and it worked for me :+1: 

Just one note, tests related to transforms just works with fresh Elastic stacks.

```
Feature: Transforms
  Define transforms in packages.
=== RUN   TestSpecCompliance/Installer_installs_the_transform_included_in_package
elastic-package has been installed.
2023/09/13 17:36:40  INFO New version is available - v0.87.0. Download from: https://github.com/elastic/elastic-package/releases/tag/v0.87.0
Error: can't install the package: could not zip-install package; API status code = 500; response body = {"statusCode":500,"error":"Internal Server Error","message":"security_exception\n\tRoot causes:\n\t\tsecurity_exception: action [indices:admin/delete] is unauthorized for service account [elastic/kibana] on indices [.metrics-transform.metadata_united_default], this action is granted by the index privileges [delete_index,manage,all]"}
exit status 1

  Scenario: Installer installs the transform included in package # features/transforms.feature:5
    Given the "transform" package is installed                   # compliance_test.go:88 -> github.com/elastic/package-spec/compliance.thePackageIsInstalled
    cannot install package "testdata/packages/transform": elastic-package failed: exit status 1
    Then there is a transform "logs-transform.metadata_united-*" # compliance_test.go:159 -> github.com/elastic/package-spec/compliance.thereIsATransform
    suite.go:476: cannot install package "testdata/packages/transform": elastic-package failed: exit status 1
=== RUN   TestSpecCompliance/Configure_aliases_for_transforms
elastic-package has been installed.
2023/09/13 17:36:46  INFO New version is available - v0.87.0. Download from: https://github.com/elastic/elastic-package/releases/tag/v0.87.0
Error: can't install the package: could not zip-install package; API status code = 500; response body = {"statusCode":500,"error":"Internal Server Error","message":"security_exception\n\tRoot causes:\n\t\tsecurity_exception: action [indices:admin/delete] is unauthorized for service account [elastic/kibana] on indices [.metrics-transform_aliases.metadata_united_default], this action is granted by the index privileges [delete_index,manage,all]"}
exit status 1

  Scenario: Configure aliases for transforms                           # features/transforms.feature:10
    Given the "transform_aliases" package is installed                 # compliance_test.go:88 -> github.com/elastic/package-spec/compliance.thePackageIsInstalled
    cannot install package "testdata/packages/transform_aliases": elastic-package failed: exit status 1
    Then there is a transform "logs-transform.metadata_united-*"       # compliance_test.go:159 -> github.com/elastic/package-spec/compliance.thereIsATransform
    And there is a transform alias "transform_aliases_transform_alias" # compliance_test.go:180 -> github.com/elastic/package-spec/compliance.thereIsATransformAlias
    suite.go:476: cannot install package "testdata/packages/transform_aliases": elastic-package failed: exit status 1

--- Failed steps:

  Scenario: Installer installs the transform included in package # features/transforms.feature:5
    Given the "transform" package is installed # features/transforms.feature:6
      Error: cannot install package "testdata/packages/transform": elastic-package failed: exit status 1

  Scenario: Configure aliases for transforms # features/transforms.feature:10
    Given the "transform_aliases" package is installed # features/transforms.feature:11
      Error: cannot install package "testdata/packages/transform_aliases": elastic-package failed: exit status 1


5 scenarios (3 passed, 2 failed)
14 steps (9 passed, 2 failed, 3 skipped)
```

File: compliance/go.mod [hunk 1]
```diff
@@ -0,0 +1,169 @@
+module github.com/elastic/package-spec/compliance
+
+go 1.21.0
+
+require (
```
[the remark this answers is not in the document]
Comment (jsoriano) [reply compliance/go.mod:5]: Good idea, I think it is possible, I will push the change.

File: .buildkite/scripts/run-installer-compliance.sh [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    $elastic_package stack up -d --version $stack_version
+    eval $($elastic_package stack shellinit)
+    cd -
+}
+
+echo "--- Start local Elastic Stack $STACK_VERSION with elastic-package"
+start_stack $STACK_VERSION
+
+echo "--- Check compliance with Package Spec $SPEC_VERSION"
+TEST_SPEC_VERSION=$SPEC_VERSION TEST_SPEC_JUNIT=report-$STACK_VERSION-$SPEC_VERSION.xml make -C compliance test
```
[the remark this answers is not in the document]
Comment (mrodm) [reply .buildkite/scripts/run-installer-compliance.sh:41]: I guess that should be a warning, but the junit report renders it as an Error? I ran the command locally and it reports 1 pending.

```
=== RUN   TestSpecCompliance/Configure_aliases_for_transforms
elastic-package has been installed.
2023/09/13 17:20:00  INFO New version is available - v0.87.0. Download from: https://github.com/elastic/elastic-package/releases/tag/v0.87.0

  Scenario: Configure aliases for transforms                           # features/transforms.feature:10
    Given the "transform_aliases" package is installed                 # compliance_test.go:88 -> github.com/elastic/package-spec/compliance.thePackageIsInstalled
    Then there is a transform "logs-transform.metadata_united-*"       # compliance_test.go:159 -> github.com/elastic/package-spec/compliance.thereIsATransform
    And there is a transform alias "transform_aliases_transform_alias" # compliance_test.go:180 -> github.com/elastic/package-spec/compliance.thereIsATransformAlias
      TODO: write pending definition

5 scenarios (4 passed, 1 pending)
14 steps (13 passed, 1 pending)
```

**Review by mrodm — APPROVED** [review]
Great! 

Not sure about the Junit report (Buildkite annotation) for that pending test. But the corresponding step does not fail, so it should be ok.

File: .buildkite/scripts/run-installer-compliance.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply .buildkite/scripts/run-installer-compliance.sh:41]: Yes, I could not find a different way to render it :shrug: It'd be great to be able to display this in a different way.

**Comment by jsoriano** [comment]
> Just one note, tests related to transforms just works with fresh Elastic stacks.

Yes, I think that this is related to EPM not removing transforms when uninstalling packages, maybe we can do something to fix this in a follow up.

**Pull request merged**